### PR TITLE
Fixed build on Linux with GCC

### PR DESCRIPTION
**Pull request by stemann**
Also, Fixed CMakeLists wrt. setting `MLX_BUILD_BENCHMARKS=OFF`.

Follow-up after #38 

Needed for https://github.com/JuliaPackaging/Yggdrasil/pull/9809

### CODE REVIEW

**Comment by stemann** [comment]
Still having some trouble with: 
* The type provided to `mlx_closure_value_and_grad_new_` in  https://github.com/ml-explore/mlx-c/blob/main/mlx/c/closure.cpp#L192 and https://github.com/ml-explore/mlx-c/blob/main/mlx/c/closure.cpp#L238
* The type provided to `mlx_closure_custom_vmap_new_` in https://github.com/ml-explore/mlx-c/blob/main/mlx/c/closure.cpp#L609, and https://github.com/ml-explore/mlx-c/blob/main/mlx/c/closure.cpp#L663

@andresy Could you take a quick peak at the error(s) in the log at https://github.com/stemann/mlx-c/actions/runs/11912756270/job/33197036071#step:8:175 ?

**Comment by stemann** [comment]
Never mind. Got it fixed: https://github.com/stemann/mlx-c/actions/runs/11916338116

**Comment by stemann** [comment]
I have removed the GitHub Actions `build` workflow from this PR, but it is available for a separate PR if you would like to get `build` CI via GitHub Actions: https://github.com/stemann/mlx-c/tree/feature/github_actions_build

File: mlx/c/array.h [hunk 1]
```diff
@@ -5,6 +5,7 @@
 
 #include "mlx/c/string.h"
 
+#include <cstdint>
```
Comment (andresy) [inline mlx/c/array.h:8]: - I did not need this one on my side, curious which GCC compiler/system are you using, and where the error is located?
- this is a C header, so includes should be C, not C++; is the error popping up in `array.cpp` or in `array.h`?

File: mlx/c/closure.cpp [hunk 2]
```diff
@@ -183,7 +183,7 @@ extern "C" mlx_closure_value_and_grad mlx_closure_value_and_grad_new_func(
             "mlx_closure_value_and_grad returned a non-zero value");
       }
       auto cpp_res =
-          std::tie(mlx_vector_array_get_(res_0), mlx_vector_array_get_(res_1));
+          std::make_pair(mlx_vector_array_get_(res_0), mlx_vector_array_get_(res_1));
```
Comment (andresy) [inline mlx/c/closure.cpp:186]: Sorry this is not really documented, but this file is auto-generated, so one should not modify it manually. Instead, modify `mlxtypes.py`

```
cpp_make_tuple = "std::make_pair" if n == 2 else "std::make_tuple"
    types.append(
        {
            "cpp": cpp_tuple + "<" + ", ".join(cpp_types) + ">",
            "alt": [cpp_tuple + "<" + ", ".join(alt_types) + ">"] + alts,
            "c_to_cpp": lambda s: cpp_make_tuple + "("
```

The file can then be generated via
```
python ../../python/closure_generator.py --implementation > closure.cpp
```

File: mlx/c/compile.h [hunk 3]
```diff
@@ -8,6 +8,8 @@
 
 #include <stdio.h>
 
+#include <cstdint>
```
Comment (andresy) [inline mlx/c/compile.h:11]: same comments as above
- this is a C header
- where is the error coming up? (I do not have it on my end)

File: mlx/c/fast.cpp [hunk 4]
```diff
@@ -3,6 +3,8 @@
 /* This file is auto-generated. Do not edit manually. */
 /*                                                    */
 
+#include <cstdarg>
```
Comment (andresy) [inline mlx/c/fast.cpp:6]: This is an auto-generated file. I do not get the error there as well, so curious to know where it occurs. For this one, and the one in `metal.cpp`, I would suggest to put additional includes in `mlx/c/private/mlx.h`

File: mlx/c/metal.cpp [hunk 5]
```diff
@@ -3,6 +3,8 @@
 /* This file is auto-generated. Do not edit manually. */
 /*                                                    */
 
+#include <cstring>
```
Comment (andresy) [inline mlx/c/metal.cpp:6]: see comment above

File: mlx/c/array.h (hunk 1 above)
Comment (stemann) [reply mlx/c/array.h:8]: GCC 11.4: https://github.com/stemann/mlx-c/actions/runs/11910841319/job/33191101517#step:4:293

Though, I'm using GCC 10 in https://github.com/JuliaPackaging/Yggdrasil/pull/9809

Comment (stemann) [reply mlx/c/array.h:8]: I might have messed up wrt. C vs C++ headers.

File: mlx/c/closure.cpp (hunk 2 above)
Comment (stemann) [reply mlx/c/closure.cpp:186]: Ah OK.

**Review by andresy — CHANGES_REQUESTED** [review]
Asking for few comments, let me know if that works for you, as I would prefer to have these added before I tag the next MLX C version

File: mlx/c/compile.h (hunk 3 above)
Comment (stemann) [reply mlx/c/compile.h:11]: It was for `uintptr_t`, it seems: https://github.com/stemann/mlx-c/actions/runs/11911185130/job/33192141400#step:5:334

File: mlx/c/array.h (hunk 1 above)
Comment (andresy) [reply mlx/c/array.h:8]: Uh ok. Definitively need a `stdint.h` there, thanks

File: mlx/c/fast.cpp (hunk 4 above)
Comment (stemann) [reply mlx/c/fast.cpp:6]: `cstdarg` was for the `va_x` functions, like `va_list`, `va_arg` etc.

File: mlx/c/metal.cpp (hunk 5 above)
Comment (stemann) [reply mlx/c/metal.cpp:6]: Noted, i.e. moving it to `mlx/c/private/mlx.h`.

File: mlx/c/fast.cpp (hunk 4 above)
Comment (stemann) [reply mlx/c/fast.cpp:6]: Noted, i.e. moving it to `mlx/c/private/mlx.h`.

File: mlx/c/closure.cpp (hunk 2 above)
Comment (stemann) [reply mlx/c/closure.cpp:186]: I added the change to the `mistypes.py` code generator, but if I run it there are a lot more changes than I would expect (I would have expected zero - keeping my manual change).

File: mlx/c/metal.cpp (hunk 5 above)
Comment (stemann) [reply mlx/c/metal.cpp]: I tried replacing it with `<string.h>`, but it failed - again complaining that there is no `strncpy`: https://github.com/stemann/mlx-c/actions/runs/11923235104

Using <cstring> (now in `mlx/c/private/mlx.h`) fixes it: https://github.com/stemann/mlx-c/actions/runs/11923278948

**Comment by stemann** [comment]
It would be great to get these changes merged in, and a new version tagged.

Your requested changes should now be implemented - except I have not run the `mlxtypes.py`-script (too many unrelated changes would occur).

File: mlx/c/array.h (hunk 1 above)
Comment (stemann) [reply mlx/c/array.h]: To be more specific, I was using:
* Locally, in a container `mcr.microsoft.com/devcontainers/cpp:1-debian-11`: GCC 10.2 (on Debian 11).
* GitHub Actions, on `"ubuntu-latest"` runner: GCC 11.4
* BinaryBuilder env. (Julia): GCC 10.2 (cross-compile env. running on an Alpine Linux (musl) - cross-compiling to various targets)

File: mlx/c/closure.cpp (hunk 2 above)
Comment (stemann) [reply mlx/c/closure.cpp]: Ah - got you message on `clang-format` now. Now executed.

**Review by andresy — APPROVED** [review]

**Pull request merged**